presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +155,203 @@ private boolean advanceToNextRowGroup()
+     * 1) Collection is not defined, because one of it's optional parent fields is null
+     * 2) Collection is null
+     * 3) Collection is defined but empty
+     * 4) Collection is defined and not empty. In this case offset value is increased by the number of elements in a collection.
+     */
+    private void calculateCollectionsOffsets(String[] path, IntList offsets, BooleanList collectionIsNull, IntList definitionLevels, IntList repetitionLevels)
+    {
+        int maxDefinitionLevel = getMaxDefinitionLevel(path);
+        int maxRepetitionLevelOfCollectionElement = getMaxRepetitionLevel(path) + 1;
+        int offset = 0;
+        offsets.add(offset);
+        for (int i = 0; i < definitionLevels.size(); i = getNextCollectionIndex(repetitionLevels, maxRepetitionLevelOfCollectionElement, i)) {
+            if (isNullField(isRequired(path), definitionLevels.get(i), maxDefinitionLevel)) {
+                collectionIsNull.add(true);
+                offsets.add(offset);
+            }
+            else if (isEmptyCollection(definitionLevels.get(i), maxDefinitionLevel)) {
+                collectionIsNull.add(false);
+                offsets.add(offset);
+            }
+            else if (isNotEmptyCollection(definitionLevels.get(i), maxDefinitionLevel)) {
+                collectionIsNull.add(false);
+                offset += countElementsOfNotEmptyCollection(repetitionLevels, maxRepetitionLevelOfCollectionElement, i + 1);
+                offsets.add(offset);
+            }
         }
-        return new RowBlock(0, blockSize, new boolean[blockSize], offsets, blocks);
+    }
+
+    private int getNextCollectionIndex(IntList repetitionLevels, int maxRepetitionLevel, int i)
+    {
+        do {
+            i++;
+        }
+        while (hasMoreElements(repetitionLevels, i) && !isNextCollectionStart(repetitionLevels, maxRepetitionLevel, i));
+        return i;
+    }
+
+    private boolean isEmptyCollection(int definitionLevel, int maxDefinitionLevel)
+    {
+        return definitionLevel == maxDefinitionLevel;
+    }
+
+    private boolean isNotEmptyCollection(int definitionLevel, int maxDefinitionLevel)
+    {
+        return definitionLevel > maxDefinitionLevel;
+    }
+
+    private int countElementsOfNotEmptyCollection(IntList repetitionLevels, int maxRepetitionLevel, int nextIndex)
+    {
+        int size = 1;
+        while (hasMoreElements(repetitionLevels, nextIndex) && !isNextCollectionStart(repetitionLevels, maxRepetitionLevel, nextIndex)) {
+            if (isElementOfCurrentCollection(repetitionLevels, maxRepetitionLevel, nextIndex)) {
+                size++;
+            }
+            nextIndex++;
+        }
+        return size;
+    }
+
+    /*
+    * Collection elements can be not primitive.
+    * For such cases elements of nested collections/structs must be filtered out
+    * */
+    private boolean isElementOfCurrentCollection(IntList repetitionLevels, int maxRepetitionLevel, int nextIndex)
```

Comment:
please inline and add a comment. We have plenty of single line methods here and it makes it harder to navigate across methods.